### PR TITLE
api: Implement QPS rate limit

### LINKED ISSUE #4500: Implement QPS rate limiter
## Development Task
refer to #4480

### PR DESCRIPTION
**Pull request by CabinfeverB**
### What problem does this PR solve?
close #4500

### What is changed and how it works?
It imports golang 'rate' and uses `rate.Limiter`which controls how frequently events are allowed to happen. It implements a "token bucket" of size b, initially full and refilled at rate r tokens per second.

And the granularity includes totality and component

### Check List

Tests

- Unit test

### Release note

```release-note
Implement QPS rate limit
```

### CODE REVIEW

File: server/self_protection.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+	totalQPSRateLimiter *rate.Limiter
+
+	enableComponentQPSLimit bool
+	componentQPSRateLimiter map[string]*rate.Limiter
+}
+
+// QPSAllow firstly check component token bucket and then check total token bucket
+// if component rate limiter doesn't allow, it won't ask total limiter
+func (rl *APIRateLimiter) QPSAllow(componentName string) bool {
```
Comment (disksing) [inline server/self_protection.go:36]: no need lock?

File: server/self_protection.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+package server
+
+import (
+	"sync"
+
+	"golang.org/x/time/rate"
+)
+
+type APIRateLimiter struct {
```
Comment (disksing) [inline server/self_protection.go:23]: Need to add some comments.

File: server/self_protection.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
+// Copyright 2021 TiKV Project Authors.
```
Comment (disksing) [inline server/self_protection.go:1]: As the main struct name is `APIRateLimiter`, how about change file name to rate limiter too?

File: pkg/ratelimiter/ratelimiter_test.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+	c.Assert(*successCount, Equals, 100)
+
+	time.Sleep(4 * time.Second) // 3+1
+	wg.Add(1)
+	CountRateLimiterHandleResult(limiter, label, successCount, failedCount, lock, wg)
+	wg.Wait()
+	c.Assert(*successCount, Equals, 101)
+}
+
+func (s *testRatelimiterSuite) TestTwoLimiters(c *C) {
```
Comment (rleungx) [inline pkg/ratelimiter/ratelimiter_test.go:149]: The test case is not finished.

File: pkg/ratelimiter/ratelimiter_test.go [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+}
+
+func initCount() (*int, *int, *sync.Mutex, *sync.WaitGroup) {
+	var lock sync.Mutex
```
Comment (rleungx) [inline pkg/ratelimiter/ratelimiter_test.go:153]: I suggest moving the logic into cases.

Comment (CabinfeverB) [reply pkg/ratelimiter/ratelimiter_test.go:166]: fixed

File: pkg/ratelimiter/ratelimiter_test.go (hunk 4 above)
Comment (CabinfeverB) [reply pkg/ratelimiter/ratelimiter_test.go:154]: updated

File: pkg/ratelimiter/concurrency_limiter.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+	mu      sync.RWMutex
+	current uint64
+	limit   uint64
+}
+
+func newConcurrencyLimiter(limit uint64) *concurrencyLimiter {
+	return &concurrencyLimiter{limit: limit}
+}
+
+func (l *concurrencyLimiter) allow() bool {
```
Comment (disksing) [inline pkg/ratelimiter/concurrency_limiter.go:31]: maybe `acquire`, `allow` looks like a read-only method.

Comment (CabinfeverB) [reply pkg/ratelimiter/concurrency_limiter.go:31]: followed [golang.org/x/time/rate](https://pkg.go.dev/golang.org/x/time/rate) Limiter

Comment (rleungx) [reply pkg/ratelimiter/concurrency_limiter.go:31]: BTW, we may need to implement the `Limit` interface if we are going to support gRPC. See https://github.com/grpc-ecosystem/go-grpc-middleware/blob/560829fc74fcf9a69b7ab01d484f8b8961dc734b/ratelimit/ratelimit.go#L11-L16

Comment (CabinfeverB) [reply pkg/ratelimiter/concurrency_limiter.go:31]: > BTW, we may need to implement the `Limit` interface if we are going to support gRPC. See https://github.com/grpc-ecosystem/go-grpc-middleware/blob/560829fc74fcf9a69b7ab01d484f8b8961dc734b/ratelimit/ratelimit.go#L11-L16

IMO, this interface may not be suitable for our scene if our limit level is service

File: pkg/ratelimit/limiter.go [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+// Limiter is a controller for the request rate.
+type Limiter struct {
+	qpsLimiter         sync.Map
+	concurrencyLimiter sync.Map
+	labelBlockList     map[string]struct{}
+}
+
+// NewLimiter returns a global limiter which can be updated in the later.
+func NewLimiter() *Limiter {
+	return &Limiter{labelBlockList: map[string]struct{}{}}
```
Comment (rleungx) [inline pkg/ratelimit/limiter.go:32]: How about using make?

File: pkg/ratelimit/limiter_test.go [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+
+	for i := 0; i < 10; i++ {
+		if i < 5 {
+			c.Assert(limiter.Allow(label), Equals, true)
+		} else {
+			c.Assert(limiter.Allow(label), Equals, false)
+		}
+	}
+	time.Sleep(time.Second)
+	limiter.Update(label, DeleteQPSLimiter())
```
Comment (rleungx) [inline pkg/ratelimit/limiter_test.go:147]: The usage is a little bit wired.

File: pkg/ratelimit/limiter.go [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+	"sync"
+
+	"golang.org/x/time/rate"
+)
+
+// Limiter is a controller for the request rate.
+type Limiter struct {
+	qpsLimiter         sync.Map
+	concurrencyLimiter sync.Map
+	labelBlockList     map[string]struct{}
```
Comment (rleungx) [inline pkg/ratelimit/limiter.go:27]: Block sounds like preventing the API been used.

**Review by disksing — APPROVED** [review]

**Review by rleungx — APPROVED** [review]

**Pull request merged**